elixir/apps/domain/lib/domain/events/replication_connection.ex

```diff
@@ -0,0 +1,314 @@
+
+    {:stream, query, [], %{state | step: :streaming}}
+  end
+
+  def handle_result(
+        [%Postgrex.Result{num_rows: 0}],
+        %__MODULE__{step: :create_publication} = state
+      ) do
+    tables =
+      state.table_subscriptions
+      |> Enum.map(fn table -> "#{state.schema}.#{table}" end)
+      |> Enum.join(",")
+
+    query = "CREATE PUBLICATION #{state.publication_name} FOR TABLE #{tables}"
+    {:query, query, %{state | step: :check_replication_slot}}
+  end
+
+  def handle_result([%Postgrex.Result{}], %__MODULE__{step: :check_replication_slot} = state) do
+    query =
+      "SELECT 1 FROM pg_replication_slots WHERE slot_name = '#{state.replication_slot_name}'"
+
+    {:query, query, %{state | step: :create_replication_slot}}
+  end
+
+  def handle_result(
+        [%Postgrex.Result{num_rows: 0}],
+        %__MODULE__{step: :create_replication_slot} = state
+      ) do
+    query =
+      "CREATE_REPLICATION_SLOT #{state.replication_slot_name} LOGICAL #{state.output_plugin} NOEXPORT_SNAPSHOT"
+
+    {:query, query, %{state | step: :start_replication_slot}}
+  end
+
+  @doc """
+    Called when we receive a message from the PostgreSQL server.
+
+    We handle the following messages:
+      1. KeepAlive: A message sent by PostgreSQL to keep the connection alive and also acknowledge
+         processed data by responding with the current WAL position.
+      2. Write: A message containing a WAL event - the actual data we are interested in.
+      3. Unknown: Any other message that we don't know how to handle - we log and ignore it.
+
+    For the KeepAlive message, we respond immediately with the current WAL position. Note: it is expected
+    that we receive many more of these messages than expected. That is because the rate at which the server
+    sends these messages scales proportionally to the number of Write messages it sends.
+
+    For the Write message, we send broadcast for each message one-by-one as we receive it. This is important
+    because the WAL stream from Postgres is ordered; if we reply to a Keepalive advancing the WAL position,
+    we should have already processed all the messages up to that point.
+  """
+
+  @impl true
+
+  def handle_data(data, state) when is_keep_alive(data) do
+    %KeepAlive{reply: reply, wal_end: wal_end} = parse(data)
+
+    wal_end = wal_end + 1
+
+    message =
+      case reply do
+        :now -> standby_status(wal_end, wal_end, wal_end, reply)
+        :later -> hold()
+      end
+
+    {:noreply, message, state}
+  end
+
+  def handle_data(data, state) when is_write(data) do
+    %Write{message: message} = parse(data)
+
+    # TODO: Telemetry: Mark start
+    message
+    |> decode_message()
+    |> handle_message(state)
+  end
+
+  def handle_data(data, state) do
+    # TODO: Sanitize the data before logging
```

Comment:
Is this something you wanted to do before deploying to prod?

Reply:
Let me remove the comment. In practice, the only sensitive data that could be logged here would have already matched insert/update/delete.